### ISSUE TITLE
List all Registered Schemes

### ISSUE THREAD

**Issue by AlbinoGeek** [fyne-io/fyne#1878]
### Is your feature request related to a problem? Please describe:
Now that we have URI work in, how do I get a list of all registered Repository by Scheme?

#### Why!?

Think: The Control Panel/Settings applet that shows you what your system's protocol associations are, and lets you change the defaults, etc.

### Is it possible to construct a solution with the existing API?
No

### Describe the solution you'd like to see:
Not sure.

---

To be specific: I know how to get the ones registered in the system, but I want to know which ones are **registered in fyne**.

**Comment by andydotxyz** [comment]
I don't follow this request. Storage repositories are not the same as registers application handlers.
What does "change the defaults" mean in this context? the toolkit only provides "file://" support - elements added by applications would not be part of a system control panel.

**Comment by AlbinoGeek** [comment]
Right, and I'm currently writing Repositories for the following schemes:

- `ftp(s)://` (self-explanatory)
- `http(s)://` as a data protocol, e.g: "browsable web" (see: [httrack](https://www.httrack.com/))
- `magnet://`
- `matrix://` (discussed in Slack)
- `steam://` (discussed in Slack)
- `union://` `file` but with dirent caching, writeback-elsewhere, and parallelism

Note: "writeback-elsewhere" is sometimes referred to as a "Union FileSystem" (Where the originating data source is considered "readonly" and writes simply go "elsewhere" (to your local cache, to another FS, etc etc.)

Because my `union` is effectively a "better `file`" for my purposes, I would replace the `file://` Repository with my `union` one.

And my application (and other applications, should scheme registration become system-wide at a later date, for example, when we talk about the context of **fynedesk**), would like to know which of these handlers is registered for a given scheme, or, for a URI that supports multiple schemes, enough information to present the user with a choice.

Examples include:

  A) You can download this file from Matrix, HTTP or FTP (only showing the options which are supported by their currently registered schemes.)
  B) Matrix scheme specification allows the embedding of documents handled by other schemes, which therefore requires my `MatrixRepository` to be able to interact with other Repository on a deeper level than the current public API allows.
  C) Many other schemes do what Matrix does, including Torrent Magnet links, ed2k (edonkey) and other P2P schemes.

For (B) I can already do this for schemes I registered by keeping their typed Repository struct reference after registration, which is hacky/clunky, but works (at least, until the global schemes thing happens as mentioned above.)

For more examples of (B, C), consider [Magnet URI "Web links to the file"](https://en.wikipedia.org/wiki/Magnet_URI_scheme#Web_links_to_the_file)

**Comment by andydotxyz** [comment]
> And my application (...), would like to know which of these handlers is registered for a given scheme, or, for a URI that supports multiple schemes, enough information to present the user with a choice.

This is supposed to be transparent. A scheme can only have 1 handler loaded, and it will "just work".
One URI cannot expose multiple schemes, so you will need to express in your code the concept that a download link can be expressed using multiple URIs.

> B) Matrix scheme specification allows the embedding of documents handled by other schemes, which therefore requires my MatrixRepository to be able to interact with other Repository on a deeper level than the current public API allows.

I don't know that I follow... The URI would be what links into other repositories. You don't (and never should) need a reference to a `Repository` instance.
If your "matrix://...." repository wishes to return a "file://..." that too will just work without any new APIs.

**Comment by AlbinoGeek** [comment]
"Download link expressed using multiple URIs" is exactly what Magnet URIs do, (and Matrix URIs)

(all examples from the above linked Magnet URI Wikipedia page)

You can specify one of each parameter (and multiple per ".N" parameter:

- `as`
  - `magnet://?as=http(s)://[web link to file(URL encoded)]`
- `xs`
  - `magnet://?xs=http(s)://<gateway-hostname>?urn:<other-uri>`
  - `magnet://?xs=dchub://[hub address]:[hub port]`
- `xt`
  - `magnet://?xt.N=[ URN of the Nth file]`
  - `magnet://?xt=urn:btih:c12fe1c06bba254a9dc9f519b335aa7c1367a88a`

Your client is expected to use "whichever protocol(s) are supported, in parallel" to speed up downloading.

---

That being said, it would appear at least one approach could be to simply ask `fyne` to assemble each URI I can use, and whichever ones return an `error`, I ignore. This may be idiomatic go as well, "just do it and check errors", v.s. checking then doing.

**Comment by andydotxyz** [comment]
> Your client is expected to use "whichever protocol(s) are supported, in parallel" to speed up downloading.

Yes I see that - but you know what protocols are supported: "file://" plus any additional schemes that you have registered.
If you can't track what your app is registering then I agree attempting to load URIs and catching the failure seems like a good appproach.

That said, the raw request of listing schemes is not something I strongly disagree with - I was more confused by the "and lets you change the defaults, etc." request. The schemes that repositories handle are not available outside of your application code.

**Comment by stuartmscott** [comment]
Being able to list the repositories could help the File Dialog surface the available repositories (maybe in the side bar) from which the user can open/save files/folders

**Comment by andydotxyz** [comment]
I think we already created another ticket for how repositories could provide favourites elements...
If we have “file”, “http” and “https” i can’t quite imagine how those 3 items would provide meaningful elements without a real favourites providers api?

**Comment by stuartmscott** [comment]
Oh for sure, but before the File Dialog can ask all the repositories for their favourites, it would need a way to access all the registered repositories.

**Comment by andydotxyz** [comment]
Hmm, that is making an assumption about how favourites will work, it could be a separate API or it might provide a dedicated lookup... let’s not use it as a use-case for this ticket until we have more of an idea on that.

**Comment by andydotxyz** [comment]
Context for not wanting to find reasons to do this until we have to:
It was agreed that repositories api would not be accessed by most devs - it’s an implementation detail to back the storage API, so exposing more info about them could be going against the design...

**Comment by AlbinoGeek** [comment]
@stuartmscott Favourites within the context of the file manager are called: "System or User Bookmarks" almost unanimously across operating system documentation. Unfortunately, these are per-machine and not per-datastore. (However, I would argue that a **new FreeDesktop specification should be created** to allow for data-aware bookmarks (e.g: Repository-favourites.)

That is something else entirely, currently, however.

---

> Yes I see that - but you know what protocols are supported: "file://" plus any additional schemes that you have registered.
> If you can't track what your app is registering then I agree attempting to load URIs and catching the failure seems like a good appproach.

I do not expect that this will always be the case, and I am already making leaps and bounds in a direction that would remove this guarantee, (that myself as the application developer actually know all the `Repository` types my application registers.) I am working towards the ability to add a `Repository` type globally (across the whole system), where other `fyne` applications would automagically benefit from its existence and registration.

This would all be abstracted away behind the storage API for any application consuming these automagically registered Repository, however, a smarter application is required to manage when and how these are actually registered (such as saving global or identity-specific Repository credentials, and making them accessible via the users' Keychain.)

My short-term end-goal for this project is that in Fynedesk, I can add a `Repository` via my automagic register, and any application I then open (including the Desktop itself) will be able to utilize this new interface,  showing these new resources in the file manager, and any other place that makes sense, seamlessly.

This is the only direction I could find that would allow the realization of cloud storage handlers, making them available to applications, while not making any of the precious/sensitive credentials associated with them to be reachable by the same program utilizing the interface.

**Comment by andydotxyz** [comment]
Can we track the discussion about favourites API in #1238 please?

There is no scope for repositories to be shared with other applications at this time.
If that is the request then fair enough, but it's not what it was introduced as.

**Comment by AlbinoGeek** [comment]
> Can we track the discussion about favourites API in #1238 please?

I did, that's the message above the horizontal rule, and I moved my comment about the spec over to that issue thread.

> There is no scope for repositories to be shared with other applications at this time.
> If that is the request then fair enough, but it's not what it was introduced as.

I've written that, which requires this, to be implemented idiomatically/"following the style of the project".

> However, a smarter application is required to manage when and how these are actually registered (such as saving global or identity-specific credentials, and making them accessible via the users' Keychain.)

This again goes back to preferences management, and the idea of global state, which I brought up with walkable preferences.

---

I full on known I am terrible when it comes to explanations without a central use-case, so here goes....

---

I cannot replace `system-*` and `gnome-*` applications with `fyne` applications until they can provide the same functionality as the application(s) they are replacing, to further the goal of `fyne` as its own distribution, without exposing or relying-on non-`fyne` graphical application.

Where should I collect issues in regards to this, if not on the `fyne` framework repository (where they would have to be implemented, as this is where the API is) itself? Perhaps something like this belongs in a proposal instead, even though it is not "definable" in the sense of a single work unit (as it relies on small changes across the API.)

These blocked application replacements so far are:

- Waiting on this feature request:
  - `gnome-online-accounts` Presents `Repository` user-wide such as Google Drive, Microsoft OneDrive, etc.
  - Tool that allows user to set their default "open with" and associations (manipulates `xdg-open` registrations)

- Waiting on loadable TTF fonts:
  - Character Map `gnome-characters`
  - Font Viewer `gnome-font-viewer`

- Waiting on global (user) or global (system) preferences:
  - Color Accuracy/Profiles `gnome-color-manager`
  - Credential Storage `gnome-keyring`
  - Radio Management `gnome-bluetooth`, "a nettool impl."
  - User Management `gnome-users`
  - Many cross-application, user-wide or system-wide preferences

- Waiting on Internationalization (3.0):
  - `system-config-language`
  - `system-config-keyboard`

---

And 100% fair if this is considered "out of scope" entirely, which is why I've been heading it up externally. Each of the blockers I have mentioned above, I can fix in a fork as a workaround/time-buying means if I have to, I'm totally okay with that.

**Comment by andydotxyz** [comment]
@AlbinoGeek having a global repository management is not as simple as preferences share.
The code to handle those repositories is the crux of it functioning.

The sorts of applications you describe as "required to replace system* and gnome*" belong in the FyneDesk project not here.
This repository is the toolkit which is used to build the apps - "fyne as a distribution" is outside the scope of this project.

I'm going to mark this as not-an-issue right now as it has got quite confusing. 
We have discussed the need for a better favourites API, to support global repository access and the desire for many system-level applications to create a full distribution - but none of these show that listing registered schemes is required at this time.

As you said you can look up specific repository based on scheme and handle based on presence/lack.
